### PR TITLE
fix: retry SetVariable on for UNAVAILABLE and RESOURCE_EXHAUSTED

### LINKED ISSUE #15549: `SetVariables` is not always retryable
With https://github.com/camunda/zeebe/pull/15230 we introduced a default service configuration that marks the `SetVariables` call as retryable on `UNAVAILABLE`, `RESOURCE_EXHAUSTED`, and `DEADLINE_EXCEEDED`:

https://github.com/camunda/zeebe/blob/f80144b951c9bd6b40d82dc10a428a1fcb4485bd/gateway-protocol-impl/src/main/resources/gateway-service-config.json#L3-L25

Since `SetVariables` is an _imperative_ command that simply sets the variables for a given element instance, it should not be retried without the user knowing. After the first `SetVariables` was processed but the response timed out with `DEADLINE_EXCEEDED`, the process instance may make progress and change state such that a retried `SetVariables` may not be correct. 

`SetVariables` stands out compared to the other methods currently marked as retryable such as `CompleteJob`, `ResolveIncident`, which are idempotent such that an accidental retry simply has no effect. It is more like `PublishMessage` or `ModifyProcessInstance` which can't always be retried.

### PR DESCRIPTION
**Pull request by berkaycanbc**
## Description
SetVariable endpoint is making changes on the state of the process instance. Therefore, it is not idempotent. If we retry setting variables in different stages of process instance, it might provide different behaviour. Because of that we shouldn't retry SetVariables on timeouts (DEADLINE_EXCEEDED) where the processing of the engine can continue, so the process instance state changes. For UNAVAILABLE and RESOURCE_EXHAUSTED, the broker will not accept further commands, so that when we retry the SetVariables command, the process instance state will stay the same.

## Related issues

closes #15549

### CODE REVIEW

**Comment by berkaycanbc** [comment]
@oleschoenburg @korthout We worked on this issue during our mob-programming session. I added our ideas in the commit message (same as in the PR description). We added you as reviewers since Nico reviewed the original PR https://github.com/camunda/zeebe/pull/15230 and Ole created the issue for this PR. Please take a look. 👀

**Review by lenaschoenburg — APPROVED** [review]

File: zeebe/gateway-protocol-impl/src/main/resources/gateway-service-config.json [hunk 1]
```diff
@@ -34,7 +33,8 @@
         {"service": "gateway_protocol.Gateway", "method": "PublishMessage"},
         {"service": "gateway_protocol.Gateway", "method": "ThrowError"},
         {"service": "gateway_protocol.Gateway", "method": "UpdateJobRetries"},
-        {"service": "gateway_protocol.Gateway", "method": "UpdateJobTimeout"}
+        {"service": "gateway_protocol.Gateway", "method": "UpdateJobTimeout"},
+        {"service": "gateway_protocol.Gateway", "method": "SetVariables"}
```
Comment (korthout) [inline zeebe/gateway-protocol-impl/src/main/resources/gateway-service-config.json:37]: 🔧 The list appears ordered alphabetically. Let's place `SetVariables` between `PublishMessage` and `ThrowError` to continue this.

**Review by korthout — APPROVED** [review]
LGTM 👍

**Comment by korthout** [comment]
@berkaycanbc we should consider whether we want to backport this

**Comment by berkaycanbc** [comment]
> @berkaycanbc we should consider whether we want to backport this

Thanks @korthout for reminding, I will apply backports for the two latest releases. 👍 

/backport

**Pull request merged**